### PR TITLE
Add flag-protected right-to-left evaluation

### PR DESCRIPTION
**Pull request by paulbrauner-da**
**TL;DR:** Right-to-left evaluation allows for a faster and simpler interpreter, but is incompatible with the semantics of Daml 2.x. Adding it behind a flag so we can switch to it in Daml 3.0.

### Problem statement

In the current language spec, expressions are evaluated from left to right, as specified by the definition of evaluation contexts:

```
 E ::= [ ]
     | E₁ e₂
     | v₁ E₂
     | …
```

That is, assuming for a moment that Daml has a `trace` function, the following term prints “AXBY” when evaluated:

```
(\x -> trace "X" (\y -> trace "Y" ()))
  (trace "A" ())
  (trace "B" ())
```

That order of evaluation unfortunately doesn't play well with the ANF transformation.

The ANF transformation transforms all applications of functions to _terms_ into applications of functions to _variables_ by introducing a let binding for each argument. This allows for a simpler CEK machine that no longer needs to push a continuation per argument for each application.

With that in mind, one would expect the term above to transform to:

```
   let f = (\x -> trace "X" (\y -> trace "Y" ()))
in let a = trace "A" ()
in let b = trace "B" ()
in f a b
```

But it can't, because the evaluation of that transformed term would print the unexpected "ABXY". This happens because we've ANF-transformed the successive applications of a curried function as if it were one single application of an uncurried function. To fix this we could perform a proper semantics-preserving ANF transformation for curried functions, but then we'd pay a steep price by introducing a closure at each intermediate application:

```
   let f = (\x -> trace "X" (\y -> trace "Y" ()))
in let a = trace "A" ()
in let fa = f a
in let b = trace "B" ()
in fa b
```

We would also be working against the first pass of the compiler which uncurries abstractions and applications.

For that reason, the interpreter currently limits the ANF transformation to unary functions and builtins. Any application of a function to more than one argument [is left untouched](https://github.com/digital-asset/daml/blob/80c844c3138e4a03b1a7b037abedf8e7ae4bfb3c/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala#L403-L404), and the CEK machine has to interpret it the "old way", by pushing a continuation per argument.

One could wonder whether the evaluation order really matters, Daml being a pure language. It unfortunately does matter, as it is observable in the presence of exceptions or non-termination.

### Solution

This PR introduces a compiler flag that enables right-to-left evaluation of expression. That is, we change the semantics of LF to:

```
 E ::= [ ]
     | e₁ E₂
     | E₁ v₂
     | …
```

With that change, the term above prints "BAXY", as does its ANF transformed version:

```
   let b = trace "B" ()
in let a = trace "A" ()
in let f = (\x -> trace "X" (\y -> trace "Y" ()))
in f a b
```

The CEK machine can then be simplified to only handle applications of functions to variables.

This is a breaking change so it is currently hidden behind a flag that will be hopefully flipped in Daml 3. In the meantime we unfortunately can't modify the CEK machine. What we can do though is carefully document via tests what changes and what doesn't when the flag is flipped.

Note that that this evaluation order is not unheard of, the Ocaml interpreter uses it for instance:

```
# (fun x -> p "X"; (fun y -> p "Y")) (p "A") (p "B");;             
BAXY- : unit = ()
```

### Implementation notes

- A good reading starting point is `AnfTest.scala` as it shows on many examples how the evaluation order is affected. The actual change to the evaluation order is implemented in `Anf.scala`.
- Synthesized expressions, like that in `translateInterfaceChoiceBody`, had to be changed with explicit let bindings to preserve their current evaluation order. Were we not to enforce that evaluation order, the change would be much more disruptive to user and break virtually all of the scala evaluation order tests.
- Many interpreter tests are affected in small ways by the new evaluation order. Even for those that aren't, we test both evaluation orders in order to catch any difference that would be later introduced. For instance, a change to the closure conversion pass could lead to different behaviors between left-to-right and right-to-left. We want to catch that difference and review it.
- I had to shuffle things around in the modified tests because many of them had method declarations interspersed between test cases. Now that the test cases are enclosed by a loop, these method definitions have to be moved outside of the loop. 
- Similarly, I had to turn some objects into classes, which had implication on the visibility of methods. I had to modify some visibility modifiers. In particular one test class is now `private[lf]`, which doesn't seem to be an issue for the test runner.
- `SemanticsEvalOrderRightToLeft.daml` would be affected if the default value of the flag was changed. I intend to modify it in a separate PR so that it tests both modes.
- We trhow a compilation error whenever someone tries to enable right-to-left evaluation in non-dev mode. We don't need a fancy error type as this should never happen by construction.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -64,12 +64,18 @@ private[lf] object Compiler {
...
+    *     speeds it up, but also changes the evaluation order or daml programs in a way that is incompatible with the
+    *     daml 2 specification. This incompatibility is observable in the presence of exceptions or non-termination.
+    */
   final case class Config(
       allowedLanguageVersions: VersionRange[LanguageVersion],
       packageValidation: PackageValidationMode,
       profiling: ProfilingMode,
       stacktracing: StackTraceMode,
       enableContractUpgrading: Boolean = false,
+      enableFullAnfTransformation: Boolean = false,
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:78]: Following our earlier discussion.... We should promote this new flag as controlling evaluation order, and refrain from even mentioning ANF in it's name.

Being able to do full ANF is the technical benefit of using right-to-left evaluation order, but from the user POV, the semantic change is the changed evaluation order.

And also perhaps we can avoid boolean blindness by taking a similar approach to the other `Mode` controlling flags. Perhaps:
```
      evalOrder: EvalOrderMode = EvalLeftToRight,
```

With:
```
  sealed abstract class EvalOrderMode extends Product with Serializable
  case object EvalLeftToRight extends EvalOrderMode // original order
  case object EvalRightToLeft extends EvalOrderMode // new order: more efficient; less suprising semantics
```

I realise it might be quite painful to change this everywhere, but it would be nice if our internal naming is consistent with the naming at which this flag gets exposed to the user.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 2]
```diff
@@ -818,31 +843,50 @@ private[lf] final class Compiler(
...
-                  },
-                ),
-                s.SCaseAlt(t.SCPDefault, env.toSEVar(mbKeyPos)),
-              ),
-            )
-        },
-      )
+      // We use a chain of let bindings to make the evaluation order of SBuildContractInfoStruct's arguments is
+      // independent from the evaluation strategy imposed by the ANF transformation.
+      let(env, s.SEValue(STypeRep(TTyCon(tmplId)))) { (typePos, env) =>
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:848]: :+1: 
Making the evaluation order of contract info components *not* be determined by function-arg evaluation order is a win.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala [hunk 3]
```diff
@@ -242,21 +281,25 @@ private[lf] object Anf {
...
-              val overApp = args.lengthCompare(b.arity) > 0
-              !overApp
-            case _ => false
-          }
-        // It's also safe to perform ANF for applications of a single argument.
-        val singleArg = args.lengthCompare(1) == 0
-        if (safeFunc || singleArg) {
-          transformMultiApp(depth, env, func, args)(transform)
+        if (enableFullAnfTransformation) {
+          transformMultiAppRightToLeft(depth, env, func, args)(transform)
```
Comment (nickchapman-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala:285]: Short note for other reviewers...

Here is the payoff.

When using the right to left evaluation order, we no longer have a concept of an expression being safe-(or-not)-for-ANF, it is always safe! And so we don't need `transformMultiAppLeftToRightSafely` which is the only constructor of (`SExpr`) `SEAppOnlyFunIsAtomic`, which is the only caller of (`Speedy`) `executeApplication`, which is a chunk of additional complexity in speedy runtime evaluation.

**Review by nickchapman-da — APPROVED** [review]

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 1 above)
Comment (paulbrauner-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala]: Done

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala [hunk 4]
```diff
@@ -40,133 +41,180 @@ import scalaz.{@@, Tag}
 
 private[lf] object Anf {
```
Comment (paulbrauner-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala:42]: Note to reviewers: the only interesting change in this file is the addition of `transformMultiAppRightToLeft`. The rest of the diff is noise due to moving some methods around.

I had to moved these methods around because I turned `Anf` into a class in order to parameterize it by the evaluation mode. Doing that forced me to move all static methods together into the object and all non-static ones together into the class. Thus most of the diff is methods being moved to another location

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.language
```
Comment (remyhaemmerle-da) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala:4]:
```suggestion
package com.daml.lf.language
```
maybe we could put it in some "features" or "devfeature" package

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala [hunk 6]
```diff
@@ -40,133 +41,180 @@ import scalaz.{@@, Tag}
...
     def n: Int = Tag.unwrap(x)
+
     def incr(m: Int): DepthA = DepthA(m + n)
   }
 
   /** `Env` contains the mapping from old to new depth, as well as the old-depth as these
     * components always travel together
     */
-  private[this] final case class Env(absMap: Map[DepthE, DepthA], oldDepth: DepthE)
+  private final case class Env(absMap: Map[DepthE, DepthA], oldDepth: DepthE)
```
[the remark this answers is not in the document]
Comment (paulbrauner-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala:77]: Because the code in `class Anf` cannot see the `private [this]` members of `object Anf`, but it can see the `private` ones.

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala (hunk 5 above)
Comment (paulbrauner-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala]: Hmm actually that file name is a mistake: I intended to call it LanguageDevConfig.scala. Done.

Would you still prefer to move it to its own package?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Anf.scala:77]: :+1:

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala]: I would prefer those definitions mention `dev` in their full name in order to make explicit they are only relevant for Dev.
Adding Dev in their name, put it inside a special package or  inside a object `LanguageDevConfig` would be enough.  
I have to choose I will go for the later. 

This is a minor remark, I let you judge.

Comment (polux) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/EvaluationOrder.scala]: Done
